eventhubsprocessor/partition_manager.py

```diff
@@ -0,0 +1,280 @@
+from bs4 import BeautifulSoup
+import requests
+from eventhubsprocessor.eh_partition_pump import EventHubPartitionPump
+from eventhubsprocessor.cancellation_token import CancellationToken
+
+class PartitionManager:
+    """
+    Manages the partition event pump execution
+    """
+    def __init__(self, host):
+        self.host = host
+        self.partition_pumps = {}
+        self.partition_ids = None
+        self.run_task = None
+        self.pump_executor = None
+        self.cancellation_token = CancellationToken()
+
+    async def get_partition_ids_async(self):
+        """
+        Returns a list of all the event hub partition ids
+        """
+        if not self.partition_ids:
+            try:
+                headers = {"Content-Type":"application/json;type=entry;charset=utf-8",
+                           "Authorization": self.host.eh_rest_auth["token"],
+                           "Host":"k8sworkshop.servicebus.windows.net"
+                          }
+
+                res = requests.get('https://{}.servicebus.windows.net/{}?timeout=60& \
+                                    api-version=2014-01'.format(self.host.eh_rest_auth["sb_name"],
+                                                                self.host.eh_rest_auth["eh_name"]),
+                                   headers=headers)
+                soup = BeautifulSoup(res.text, "lxml-xml") # process xml response
+                self.partition_ids = [pid.text for pid in soup.find("PartitionIds")]
+            except Exception as err:
+                raise Exception("failed to get partition ids", err)
+
+        return self.partition_ids
+
+    async def start_async(self):
+        """
+        Intializes the partition checkpoint and lease store and then calls run async.
+        """
+        if self.run_task:
+            raise Exception("A PartitionManager cannot be started multiple times.")
+
+        partition_count = await self.initialize_stores_async()
+        logging.info("%s PartitionCount: %s", self.host.guid, partition_count)
+        self.pump_executor = concurrent.futures.ThreadPoolExecutor(max_workers=partition_count)
+
+        self.run_task = await self.run_async()
+
+    async def stop_async(self):
+        """
+        Terminiates the partition manger.
+        """
+        self.cancellation_token.cancel()
+        local_run_task = self.run_task
+        if local_run_task:
+            await local_run_task()
+
+    async def run_async(self):
+        """
+        Starts the run loop and manages exceptions and cleanup
+        """
+        try:
+            await self.run_loop_async(self.cancellation_token)
+        except Exception as err:
+            logging.error("Run loop failed %s", repr(err))
+        try:
+            logging.info("Shutting down all pumps %s", self.host.guid)
+            await self.remove_all_pumps_async("Shutdown")
+        except Exception as err:
+            raise Exception("failed to remove all pumps", repr(err))
```

Comment:
I am adding EventHubException in eventhubs package. Need to use that instead of the generic Exception.

Reply:
Will update it. Let me know when this is ready!

Reply:
We don't have to update it in this PR. The event hub error is part of the send implementation and I am not sure when it will be ready for merge. We can fix exceptions later.